commons/src/build_output.rs

```diff
@@ -0,0 +1,716 @@
+        }
+    }
+
+    impl Display for ErrorInfo {
+        fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
+            f.write_str(error(self).as_str())
+        }
+    }
+
+    /// Need feedback on this interface
+    ///
+    /// Should it take fields or a struct
+    /// Struct is more robust to change but extra boilerplate
+    /// If the struct is always needed, this should perhaps be an associated function
+    #[must_use]
+    pub fn error(info: &ErrorInfo) -> String {
+        let ErrorInfo {
+            header,
+            body,
+            url,
+            debug_details,
+        } = info;
+
+        let body = look_at_me(ERROR_COLOR, "ERROR:", header, body, url);
+        if let Some(details) = debug_details {
+            format!("{body}\n\nDebug information: {details}")
+        } else {
+            body
+        }
+    }
+
+    /// Need feedback on this interface
+    ///
+    /// Should it have a dedicated struct like `ErrorInfo` or be a function?
+    /// Do we want to bundle warnings together and emit them at the end? (I think so, but it's out of current scope)
+    pub fn warning(header: impl AsRef<str>, body: impl AsRef<str>, url: &Option<String>) -> String {
+        let header = header.as_ref();
+        let body = body.as_ref();
+
+        look_at_me(WARNING_COLOR, "WARNING:", header, body, url)
+    }
+
+    /// Need feedback on this interface
+    ///
+    /// Same questions as `warning`
+    pub fn important(
+        header: impl AsRef<str>,
+        body: impl AsRef<str>,
+        url: &Option<String>,
+    ) -> String {
+        let header = header.as_ref();
+        let body = body.as_ref();
+
+        look_at_me(IMPORTANT_COLOR, "", header, body, url)
+    }
+
+    fn help_url(body: impl AsRef<str>, url: &Option<String>) -> String {
+        let body = body.as_ref();
+
+        if let Some(url) = url {
+            let url = colorize(URL_COLOR, url);
+
+            formatdoc! {"
+            {body}
+
+            For more information, refer to the following documentation:
+            {url}
+        "}
+        } else {
+            body.to_string()
+        }
+    }
+
+    /// Helper method that adds a bang i.e. `!` before strings
+    fn bangify(body: impl AsRef<str>) -> String {
+        body.as_ref()
+            .split('\n')
+            .map(|section| format!("! {section}"))
+            .collect::<Vec<String>>()
+            .join("\n")
+    }
+
+    /// Colorizes a body while preserving existing color/reset combinations and clearing before newlines
+    ///
+    /// Colors with newlines are a problem since the contents stream to git which prepends `remote:` before the `libcnb_test`
+    /// if we don't clear, then we will colorize output that isn't ours.
+    ///
+    /// Explicitly uncolored output is handled by a hacky process of treating two color clears as a special case
+    pub(crate) fn colorize(color: &str, body: impl AsRef<str>) -> String {
```

Comment:
All this search and replace is a bit worrying. There are many ways this could go wrong. I'd drop all these `replace` calls. If someone wants to ensure some text is uncolored, they can use `/x1b[39m`, which specifies default foreground color. Then, wrapping with red won't override it.

Reply:
Your comments on escape codes Are really helpful. This is my main resource https://sweworld.net/linux/terminal/escape_code/ 

This is a great suggestion, and it clears my hack:

>  \x1b[1;39m

But it makes me realize I guess I don't know what \x1b[0m is actually doing. I thought it was resetting things, but I don't know what would be different between that and "Default"

Reply:
Updated to remove the replace hack for NOCOLOR (which turns out wasn't even used). I'm still curious about the difference though (if you know)

Regarding `replace` usage, I couldn't find any off-the-shelf colorization libraries that handled the color-within-a-color. I'm game to update this to another approach. One idea I had was using `nom` to parse the incoming text and adjust it somehow. Or maybe I could try to grab some time with burnt sushi (who has significant experience with colorization). But I might put that off until we work to formalize the implementation. 

FWIW there's not a ton of tests in this new code, but this colorize function is tested. This method and the associated constants are all private and cannot be accessed except from within the module.

Reply:
You are right - `\x1b[0m` is a full reset. It clears all active styles - foreground color, background color, bold, dim, underline, blink (and others I'm probably forgetting). 

`\x1b[39m` sets the foreground color to the default color. I don't think `\x1b[39m` will fix all problems though. For example:

```
\x1b[31m]This is red\x1b[39mThis is default\x1b[0mThis is also default, even though we wanted to wrap it in red.\x1b[0m
```

With regard to `replace` usage, I was hoping to drop it completely rather than find something new. Maybe `colorize` shouldn't support color-in-color text - after all, this isn't really a thing ANSI escape codes support. ANSI escape codes don't have a tree-like structure by nature.

Reply:
> Maybe colorize shouldn't support color-in-color text

If we can't have style-in-a-style then it would have the effect of forcing "warning", "error", and "important" to all be all one color. I think it's nice to have this ability, take the `! Help`  output for example, it's colorized to draw your attention, but then there are values inside of it. 

Right now there are no control codes that go into `colorize` that we don't own so I have a high degree of confidence in its implementation. One reason that's true is that currently, other build tools (like bundle install) are not being detected as a TTY. But we've not talked about that before:

From CNB execution https://github.com/heroku/buildpacks-ruby/actions/runs/5523980642/jobs/10075643688?pr=155

![image](https://github.com/heroku/buildpacks-ruby/assets/59744/94a9f124-26e0-478b-8e2f-06a73e9ef6e6)

Running `bundle install` locally: 

![image](https://github.com/heroku/buildpacks-ruby/assets/59744/333744cf-61e2-4eb8-9688-14189d8cccf8)

Notice the green is present locally but not on CNB deploys.

I think this CNB behavior matches classic behavior, but it does ask the question: What output do we want? Should our subcommands have native colorization or be plain?

From an implementation standpoint, I think supporting colorized subcommands makes life harder (but not impossible). For this PR I don't want to change this behavior, but I do want to understand/explore our idealized outcome.